alert_L module for handling oob alerts

nchatrad opened the pull request:
splitting the previous PR 135

akky16 committed:
dt-bindings: misc: apml-alertl: Add APML Alert_L bindings

-Document device-tree bindings for APML Alert_L driver.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Signed-off-by: sathya priya kumar <SathyaPriya.K@amd.com>
(cherry picked from commit 0f54429b0d8f9a2e935fa7c300601444ba54ea15)

akky16 committed:
misc: amd-apml: Add support for AMD APML Alert_L platform driver

Processors from AMD provide APML ALERT_L for BMC users to
monitor events.
APML Alert_L is asserted in multiple events,
- Machine Check Exception occurs within the system
- The processor alerts the SBI on system fatal error event
- Set by hardware as a result of a 0x71/0x72/0x73 command completion
- Set by firmware to indicate the completion of a mailbox operation
- High/Low Temperature Alert

APML Alert_L module define uevents to notify registered userspace processes
of the alert event.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Signed-off-by: sathya priya kumar <SathyaPriya.K@amd.com>

akky16 committed:
misc: amd-apml: Create common header for sbtsi device

Move out the sbtsi device structure to a common header file and
add device matching helper function for I2C and I3C buses.

This is required to add support for the APML Alert_L module which
needs access to sbtsi device struct and device matching capability.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>

mahkurap commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-congo.dts:
```diff
@@ -894,6 +894,17 @@
+		status = "okay";
+		gpios = <&ltpi0_gpio 20 GPIO_ACTIVE_LOW>;
+		sbrmi = <&sbrmi_p0_iod0>;
+		sbtsi = <&sbtsi_p0_iod0>;
```
why is &sbtsi_p0_iod1 missing for Congo platforms?

sathyapk replied:
Updated &sbtsi_p0_iod1 for Congo platform

mahkurap replied:
we may not need iod1. at the time of initial review BMC was reading the iod1 temperature as well. That is why I thought we need iod1 for alert_l also.  Can you please check the requirements with the team and add only if needed? apologies for the to and fro on this.

sathyapk replied:
Sure, will confirm and update. thanks.

sathyapk replied:
CF team confirmed IOD1 is not applicable in mission mode. Hence removed IOD1 related changes in DTS.

mahkurap commented on Documentation/devicetree/bindings/misc/amd,apml-alertl.yaml:
```diff
@@ -0,0 +1,50 @@
+required:
+  - compatible
+  - gpio
+  - sbrmi
```
please add more description for these tags. Like what is the input, parameters.. etc.

sathyapk replied:
Done

mahkurap commented on Documentation/devicetree/bindings/misc/amd,apml-alertl.yaml:
```diff
@@ -0,0 +1,50 @@
+  - compatible
+  - gpio
+  - sbrmi
+  - sbtsi
```
I saw that in Congo there is only iod0 is mentioned, and in morocco iod1 also mentioned. Please explain more here on the parameters for this tag.

mahkurap commented on include/uapi/linux/amd-apml.h:
```diff
@@ -7,11 +7,21 @@
+#define BIT(n) (1U << (n))
+#endif
+
+enum apml_ras_alert_src {
```
why are we calling this _ras_ alert?

sathyapk replied:
Renamed to apml_alert_src

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+		.of_match_table = of_match_ptr(apml_alertl_dt_ids),
+	},
+	.probe		= apml_alertl_probe,
+	.remove		= alert_remove,
```
for uniformity, change alret_remove to apml_alaertl_remove()

sathyapk replied:
Done

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+	return 0;
+}
+
+static int alert_remove(struct platform_device *pdev)
```
change name to apml_alertl_remove()

sathyapk replied:
Done

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+
+	snprintf(sock, sizeof(sock), "Socket=0x%x", soc_die_num);
+	snprintf(src, sizeof(src), "Source=0x%x", alert_src);
+	pr_err("apml_alertl:Sock:0x%x Src:0x%x\n", soc_die_num, alert_src);
```
do you mean to have pr_err() here? should it be pr_dbg()?

sathyapk replied:
yes, that's correct. 
moved all the logging to dev_dbg/dev_err

mahkurap reviewed (changes requested):
Please add apml_alert dts changes in Kenya and Nigeria device trees as well.
Also, change congo to Congo, morcco to Morocco in one of the commit message.

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+	 * 0x3c, 0x4c	-> Socket 0, die 0,
+	 * 0x44		-> Socket 0, die 1,
+	 * 0x38, 0x48	-> Socket 1, die 0,
+	 * 0x4c		-> Socket 1, die 1.
```
Typo ? 0x45?

sathyapk replied:
Yes, corrected

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+	return ret;
+}
+
+static u8 static_addr_to_socket(u8 static_addr)
```
Is this rule is valid incase 2X1P configuration boot .

akky16 replied:
As per my understanding
2X1P and 2P can be used interchangeably.
As the PID and static address are related to Socket and Die ID which are stored in CSR registers(RO)
This should be valid.

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,362 @@
+	return 0;
+}
+
+static irqreturn_t alert_l_irq_thread_handler(int irq, void *dev_id)
```
Error Handling , general: Since this is alert handling, it's better to follow a best-effort approach for error handling?

sathyapk replied:
Thanks for the input.
Update the same

ojayanth commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-congo.dts:
```diff
@@ -894,6 +894,17 @@
 #endif
 };
 
+/ {
```
Commit message : start with ARM64:dts:aspeed:

sathyapk replied:
Done

ojayanth commented on Documentation/devicetree/bindings/misc/amd,apml-alertl.yaml:
```diff
@@ -0,0 +1,50 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
```
Commit message Please align with upstream style : dt-bindings: misc: apml-alert

sathyapk replied:
Done

ojayanth commented:
Please ensure your commit message style aligns with the conventions used in the relevant upstream directory

sathyapk committed:
arm64: dts: aspeed: Alert_L changes for Morocco & Congo

-Add alertl node for P1 and P2 host processor in Morocco platform.
-Add alertl node for P1 host processor in Congo platform.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Signed-off-by: sathya priya kumar <SathyaPriya.K@amd.com>

sathyapk committed:
arm64: dts: aspeed: Alert_L changes for Kenya & Nigeria

-Add alertl node for P1 and P2 host processor in Nigeria platform.
-Add alertl node for P1 host processor in Kenya platform.

Reviewed-by: Naveen Krishna Chatradhi <naveenkrishna.chatradhi@amd.com>
Signed-off-by: Akshay Gupta <akshay.gupta@amd.com>
Signed-off-by: sathya priya kumar <SathyaPriya.K@amd.com>

mahkurap commented on Documentation/devicetree/bindings/misc/amd,apml-alertl.yaml:
```diff
@@ -0,0 +1,63 @@
+  gpio:
+    maxItems: 1
+    description: |
+	GPIO specifier for APML Alert_L line. Specify GPIO controller, GPIO pin
```
nit: white space?

sathyapk replied:
will update, thanks

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+MODULE_ALIAS("apml_alertl:" DRIVER_NAME);
+
+/* Read TSI Status register to identify the RAS error */
```
nit: correct the comment

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	snprintf(sock, sizeof(sock), "Socket=0x%x", soc_die_num);
+	snprintf(src, sizeof(src), "Source=0x%x", alert_src);
+
+	dev_err(dev, "Sending uevent: Sock:0x%x Src:0x%x\n",
```
dev_err will overwhelm the journalctl/dmesg. change it to dev_dbg(). should be able to enable it at runtime.

sathyapk replied:
Udpated

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		return IRQ_NONE;
+
+	for (i = 0; i < oob_adata->num_of_tsi_devs; i++) {
+		rt_src = 0;
```
nit: can be removed. no effect statement

sathyapk replied:
Done

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		rt_src = 0;
+		/* Check for NULL pointer */
+		if (IS_ERR_OR_NULL(oob_adata->tsi_dev[i])) {
+			dev_err(dev, "TSI device does not exist\n");
```
Add "i", and any other data to the log, will be more helpful during debug.

sathyapk replied:
Sure, added

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	msg.data_in.reg_in[REG_OFF_INDEX] = TSI_STATUS_REG;
+
+	mutex_lock(&oob_adata->tsi_dev[soc_die_num]->lock);
+	ret = regmap_read(oob_adata->tsi_dev[soc_die_num]->regmap,
```
if the regmap_read() fails for any reason, does *temp_status set to zero?

sathyapk replied:
temp_status is initialized to zero

mahkurap replied:
temp_status is initialized outside of the for loop(). In case of failures, we are checking the previously returned by the previous iteration.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		/* Check for TSI alert */
+		ret = tsi_alert_check(oob_adata, &temp_status, i);
+		if (ret < 0)
+			dev_err(dev, "Failed to read TSI status register\n");
```
Additional data in the error logging will be of great help during debugging.  Add more context to the logging.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		if (ret < 0)
+			dev_err(dev, "Failed to read TSI status register\n");
+
+		if (!temp_status)
```
temp_status is not guaranteed to set to zero incase of failures. add this check to the previous check.

sathyapk replied:
temp_status initialized to zero. On failure, temp_status remains zero

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		rt_src = (temp_status << 24);
+		static_addr = oob_adata->tsi_dev[i]->dev_static_addr;
+
+		ret = send_uevent(static_addr, rt_src, dev);
```
rt_src can be replaced with (temp_status << 24). rt_src is not used anywhere else. Same with the variable static_addr.

sathyapk replied:
Done

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+		ret = send_uevent(static_addr, rt_src, dev);
+		if (ret)
+			dev_err(dev, "Failed to send TSI uevent Err: %d\n", ret);
```
Add (temp_status << 24), and oob_adata->tsi_dev[i]->dev_static_addr to logging.

sathyapk replied:
added more info

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	struct device *dev = oob_adata->dev;
+	unsigned int ras_status = 0;
+	unsigned int temp_status = 0;
+	u32 rt_src = 0;
```
optimization: rt_src, and static_addr variables are not needed/

sathyapk replied:
Done

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+{
+	struct apml_alertl_data *oob_adata = dev_id;
+	struct apml_message msg = { 0 };
+	struct device *dev = oob_adata->dev;
```
can the dev be null?

sathyapk replied:
no, dev cannot be null

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+static irqreturn_t alert_l_irq_thread_handler(int irq, void *dev_id)
+{
+	struct apml_alertl_data *oob_adata = dev_id;
```
no need to cast?

sathyapk replied:
not necessarily. i have updated. Thanks

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+		/* Check for TSI alert */
+		ret = tsi_alert_check(oob_adata, &temp_status, i);
+		if (ret < 0)
```
tsi_alrert_check() returns 0 if the input parameters are invalid. Change this condition to "<=" or change the tsi_alert_check() to return a negative integer incase of input error validation.

sathyapk replied:
not required with new changes

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	int ret;
+
+	if (!oob_adata->rmi_dev[soc_die_num] || !oob_adata->rmi_dev[soc_die_num]->regmap)
+		return 0;
```
change to a negative value. 0 return code is  ==> success

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	int ret;
+
+	if (!oob_adata->tsi_dev[soc_die_num] || !oob_adata->tsi_dev[soc_die_num]->regmap)
+		return 0;
```
return negative integer in case of error.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+	soc_die_num = static_addr_to_socket(static_address);
+	if (soc_die_num == 0xFF) {
+		dev_err(dev, "Device static address not valid\n");
```
Log static_address in case of error

sathyapk replied:
i dont think it's needed, as we are retuning 0xFF static_addr_to_socket()

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	for (i = 0; i < oob_adata->num_of_rmi_devs; i++) {
+		rt_src = 0;
```
Please move the tsi/rmi check and event handling to different routines keeping this routine as a overall handler if possible.

sathyapk replied:
Yes, done!

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	}
+
+	for (i = 0; i < oob_adata->num_of_rmi_devs; i++) {
+		rt_src = 0;
```
rt_src ==> can be removed.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		rt_src = 0;
+		/* Check for NULL pointer */
+		if (IS_ERR_OR_NULL(oob_adata->rmi_dev[i])) {
+			dev_err(dev, "RMI device doesn't exist \n");
```
please add more context to the error logging. Add requried data to the logging.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		/* Check for NULL pointer */
+		if (IS_ERR_OR_NULL(oob_adata->tsi_dev[i])) {
+			dev_err(dev, "TSI device does not exist\n");
+			return IRQ_NONE;
```
should this be changed to continue? the RMI related events might never get triggered/serviced. please check.

sathyapk replied:
Yes, it should be. Updated

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		/* Check for NULL pointer */
+		if (IS_ERR_OR_NULL(oob_adata->rmi_dev[i])) {
+			dev_err(dev, "RMI device doesn't exist \n");
+			return IRQ_NONE;
```
dont we need to check for all the devices before returning?

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		if (ret < 0)
+			dev_err(dev, "Failed to read RAS status register\n");
+
+		if (!ras_status)
```
similar issue like int the case of tsi. See if the tsi_alert_check/rmi_alert_check function can be passed as a parameter along with context to a common handler routine. Lot of this duplicate code can be avoided. my 2 cents.

sathyapk replied:
This alert check for rmi/tsi is done as part of respective handler routine.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		if (!ras_status)
+			continue;
+
+		rt_src = ras_status;
```
rt_src, and static_addr variables can be avoided.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+		ret = send_uevent(static_addr, rt_src, dev);
+		if (ret)
+			dev_err(dev, "Failed to send RMI uevent Err: %d\n", ret);
```
pls add more context to the logging. in this case rt_src, and static_addr.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+		 /* Clear the RMI Status and RAS Status register */
+		if (!oob_adata->rmi_dev[i] || !oob_adata->rmi_dev[i]->regmap) {
+			dev_err(dev, "Error accessing the RMI device");
```
add more context to the logging. "i" in this case

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+			return IRQ_NONE;
+		}
+
+		if (ras_status) {
```
pls add comments on what is being done if there is a change in the ras_status?

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+					   msg.data_in.reg_in[REG_OFF_INDEX],
+					   ras_status);
+			if (ret < 0)
+				dev_err(dev, "Failed to clear RMI status register\n");
```
add more info in the logging on for which device clearing failed. regmap_write() can it fail?

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+			if (ret < 0)
+				dev_err(dev, "Failed to clear RMI status register\n");
+
+			msg.data_in.reg_in[REG_OFF_INDEX] = STATUS_REG;
```
do we need update this again? line #192 did the same thing.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+					   RAS_ALERT_ASYNC);
+			mutex_unlock(&oob_adata->rmi_dev[i]->lock);
+			if (ret < 0) {
+				dev_err(dev, "Failed to clear RAS status register\n");
```
have a different string for logging to distinguish between this line and the above error message. also add more context.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+			mutex_unlock(&oob_adata->rmi_dev[i]->lock);
+			if (ret < 0) {
+				dev_err(dev, "Failed to clear RAS status register\n");
+				return IRQ_NONE;
```
1. Shouldn't this be changed to IRQ_HANDLED? Please check the guidelines on what all error conditions should be treated for IRQ_NONE.
2. shouldn't this be a continue to service the alerts on other RMI devices.  Otherwise the device last in the loop could be starved.

sathyapk replied:
Yes, right. it should continue to serve other RMI device. Updated.

mahkurap commented on include/uapi/linux/amd-apml.h:
```diff
@@ -7,11 +7,26 @@
+ * SBTSI 0x02 Temperature alert status bits(3 & 4)
  */
-#define USR_SIGNAL	44
+enum apml_alert_src {
```
nit: indent the BIT() definitions to the same column.

mahkurap commented on include/uapi/linux/amd-apml.h:
```diff
@@ -7,11 +7,26 @@
+	APML_MCA_ALERT		= BIT(3),
+	APML_DRAM_CECC_ALERT	= BIT(4),
+	APML_PCIE_ALERT		= BIT(5),
+	APML_CPU_SHUTDOWN	= BIT(6),
```
nit: change to APML_CPU_SHUTDOWN_ALERT

sathyapk replied:
Good catch. Updated.

mahkurap commented on include/uapi/linux/amd-apml.h:
```diff
@@ -7,11 +7,26 @@
+	APML_DRAM_CECC_ALERT	= BIT(4),
+	APML_PCIE_ALERT		= BIT(5),
+	APML_CPU_SHUTDOWN	= BIT(6),
+	APML_TEMP_LOW_ALERT	= BIT(27),
```
Any reason why bits 27 and 28 are used? They are in the middle of a higher order byte. Just curious. Please add a coment to clarify.

sathyapk replied:
udpated the comment section

mahkurap commented on include/uapi/linux/amd-apml.h:
```diff
@@ -7,11 +7,26 @@
 
 #include <linux/types.h>
 
+#ifndef BIT
```
why to redefine BIT(), Linux kernel has a rich set of macros defined for bit related operations, I thought.

sathyapk replied:
Defining BIT macro if not already defined since it's uapi header

mahkurap commented on drivers/misc/amd-apml/apml_alertl.h:
```diff
@@ -0,0 +1,25 @@
+#define _AMD_APML_ALERT_L__
+
+#include "sbrmi-common.h"
+
```
no need to include the "sbtsi-common.h" for struct apml_sbtsi_device?

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+#include <linux/regmap.h>
+#include <linux/platform_device.h>
+
+#include "sbrmi-common.h"
```
already included in the apml_alert.h

sathyapk replied:
Yes, removed in this file

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+#include "sbrmi-common.h"
+#include "apml_alertl.h"
+#include "sbtsi-common.h"
```
tsi-common.h can be removed if this also can be moved to apml_alert.h.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+#define RAS_STATUS_REG		0x4C
+#define STATUS_REG		0x2
+#define TSI_STATUS_REG		0x2
+#define RAS_ALERT_STATUS	BIT(1)
```
should the user space application need to be aware of the type of RAS_ALERT? If yes, please define a new header file for the user space applications, and move these definitions to that new file.

sathyapk replied:
yes, The alert bit information for RAS and temp is defined in amd-apml.h for the userspace apps

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	 * 0x45		-> Socket 1, die 1.
+	 */
+	switch (static_addr) {
+	case 0x3c:
```
please define macros for the static addresses, will improve the readability of the function.

sathyapk replied:
Thanks for suggestion. updated!

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+static int send_uevent(u8 static_address, u32 alert_src, struct device *dev)
+{
+	u8 soc_die_num = 0;
```
no need to initialize to 0

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+	dev_err(dev, "Sending uevent: Sock:0x%x Src:0x%x\n",
+		soc_die_num, alert_src);
+	kobject_uevent_env(&dev->kobj, KOBJ_CHANGE, alert_source);
```
How does the user space application know the structure of the "alert_source" to interpet the data? Is there a interface header file defined?

sathyapk replied:
apml_library is the interface to the alert_l driver. user space application should call apml library api's to interpret the data

sathyapk replied:
Yes, It is possible. The event structure is define in header file [drivers/misc/amd-apml/apml_alertl.h]

> [AMD Official Use Only - AMD Internal Distribution Only] How does the apml_library know the structure of the event? Is there a header file defined for it? Imagine a case where in the user wants to use the events directly with out the library interface. Is it not possible? Library also runs in the user space. From: sathyapk ***@***.***> Sent: Wednesday, August 13, 2025 2:39 AM To: AMDESE/linux-aspeed ***@***.***> Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***> Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143) Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding. @sathyapk commented on this pull request.
> ________________________________ In drivers/misc/amd-apml/apml_alertl.c<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272358558)>:
> + char sock[MAX_SOC_LEN];
> + char src[MAX_ERR_LEN]; + char *alert_source[] = { sock, src, NULL }; + + soc_die_num = static_addr_to_socket(static_address); + if (soc_die_num == 0xFF) { + dev_err(dev, "Device static address not valid\n"); + return -ENODEV; + } + + snprintf(sock, sizeof(sock), "Socket=0x%x", soc_die_num); + snprintf(src, sizeof(src), "Source=0x%x", alert_src); + + dev_err(dev, "Sending uevent: Sock:0x%x Src:0x%x\n", + soc_die_num, alert_src); + kobject_uevent_env(&dev->kobj, KOBJ_CHANGE, alert_source); apml_library is the interface to the alert_l driver. user space application should call apml library api's to interpret the data — Reply to this email directly, view it on GitHub<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272358558)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJVI6KJ4DOPVIEDGKOT3NLTRPAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMYTSNZXGM>. You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	int ret, i;
+
+	if (!oob_adata || atomic_read(&oob_adata->removed))
+		return IRQ_NONE;
```
Log the error message?

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		return NULL;
+
+	if (strcmp(phandle_name, "sbrmi") == 0) {
+		dev = bus_find_device(&i3c_bus_type, NULL, d_node, sbrmi_match_i3c);
```
where is this function sbrmi_match_i3c() defined? Did not find it in this file.

sathyapk replied:
it is part of sbrmi.c

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+
+	of_node_put(d_node);
+	apml_dev = dev_get_drvdata(dev);
+	if (IS_ERR_OR_NULL(apml_dev))
```
add an error log message.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		return 0;
+	msg.data_in.reg_in[REG_OFF_INDEX] = TSI_STATUS_REG;
+
+	mutex_lock(&oob_adata->tsi_dev[soc_die_num]->lock);
```
mutex_lock() can sleep.  As far as I know it is not safe to use mutex_lock() in an irq context. check for any alternatives.

sathyapk replied:
Lock is called inthe bottom half of the irq, mutex lock can be invoked in bottom half.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	tsi_dev = devm_kzalloc(dev, oob_alert->num_of_tsi_devs * sizeof(struct apml_sbtsi_device),
+			       GFP_KERNEL);
+	if (!tsi_dev)
+		return -ENOMEM;
```
missing free of oob_alert, and rmi_dev

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	struct apml_alertl_data *oob_alert;
+	struct gpio_desc *alertl_gpiod;
+	u32 irq_num;
+	u32 num_dev = 0;
```
no need to initialize to zero. num_dev can be eliminated.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	for (i = 0; i < oob_alert->num_of_rmi_devs; i++) {
+		rmi_dev[i] = get_apml_dev_byphandle(pdev->dev.of_node, "sbrmi", i);
+		if (!rmi_dev[i]) {
+			dev_err(dev, "Error getting APML SBRMI device\n");
```
guideline suggests not to log any error to avoid spam. I am not sure though.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+		rmi_dev[i] = get_apml_dev_byphandle(pdev->dev.of_node, "sbrmi", i);
+		if (!rmi_dev[i]) {
+			dev_err(dev, "Error getting APML SBRMI device\n");
+			return -EPROBE_DEFER;
```
if we are expecting the probe to be deferred and return, TSI device could never get probed? Not sure. pls check. I dont know how to continue either..

sathyapk replied:
Driver returns -ENODEV, as all the tsi/rmi for all the sockets need to be available before alert_l driver to function.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	struct gpio_desc *alertl_gpiod;
+	u32 irq_num;
+	u32 num_dev = 0;
+	int ret = 0;
```
no need to initialize to 0.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	u32 irq_num;
+	u32 num_dev = 0;
+	int ret = 0;
+	int i = 0;
```
no need to initialize to 0.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	irq_num = ret;
+	dev_dbg(dev, "Register IRQ:%u\n", irq_num);
+	/*
+	 * TODO: naming can be updated for the irq on
```
is this still a TODO, or just a miss?

sathyapk replied:
Thanks for pointing out. Done!

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	ret = devm_request_threaded_irq(dev, irq_num,
+					NULL,
+					(void *)alert_l_irq_thread_handler,
+					IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
```
can the IRQ attributes be device tree parameters? Makes the module more configurable.

sathyapk replied:
This can be done partly.

mahkurap replied:
I saw they are already described. pls ignore my comment.

mahkurap commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+	 * over I3C. We can add property in dts to identify the bus type
+	 */
+
+	for (i = 0; i < oob_alert->num_of_rmi_devs; i++) {
```
Where are the mutex locks initialized for RMI and TSI devices?

sathyapk replied:
The mutex locks are initialized in respective RMI and TSI drivers.

mahkurap reviewed (changes requested):
Please check comments.

ojayanth commented on drivers/misc/amd-apml/Kconfig:
```diff
@@ -24,3 +24,14 @@ config APML_SBTSI
 
       This driver can also be built as a module. If so, the module will
       be called apml_sbtsi.
+
```
NIT: commit message "driver:amd-apml: Add Alert_L support for event monitoring"

sathyapk replied:
Done

ojayanth commented on drivers/misc/amd-apml/Kconfig:
```diff
@@ -24,3 +24,14 @@ config APML_SBTSI
+    default n
+    help
+      If you say yes here you get support for emulated alertl
+      interface on AMD SoCs with APML interface connected to a BMC device.
```
NIT: 
      Alert_L may be asserted for various conditions, including machine check
      exceptions, system fatal errors, APML command completions, and thermal
      alerts

sathyapk replied:
thanks, updated!

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * alert_l.c - Alert_l driver for AMD APML devices
```
fix file name apml_alertl.c

sathyapk replied:
done, thanks

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+/*
+ * alert_l.c - Alert_l driver for AMD APML devices
+ *
+ * Copyright (C) 2022-2023 Advanced Micro Devices, Inc.
```
2025-2026?

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+ * Copyright (C) 2022-2023 Advanced Micro Devices, Inc.
+ */
+
+#include <linux/debugfs.h>
```
NIT: reorder based Kernel coding style 
Ordering Logic:
Core kernel headers: module.h, init.h
Platform and device-related: platform_device.h, interrupt.h, regmap.h
Peripheral interfaces: gpio, debugfs, i3c
Device tree support: of.h, of_gpio.h
Local headers: "sbrmi-common.h" comes last

#include <linux/module.h>
#include <linux/init.h>
#include <linux/platform_device.h>
#include <linux/interrupt.h>
#include <linux/regmap.h>
#include <linux/gpio/consumer.h>
#include <linux/debugfs.h>
#include <linux/i3c/device.h>
#include <linux/of.h>
#include <linux/of_gpio.h>

sathyapk replied:
Thanks for the info. We have been following the alphabetical order of arranging headers.

ojayanth commented on drivers/misc/amd-apml/apml_alertl.c:
```diff
@@ -0,0 +1,390 @@
+#define DRIVER_NAME "apml_alertl"
+
+#define RAS_STATUS_REG		0x4C
+#define STATUS_REG		0x2
```
NIT be constient on formating.

mahkurap commented:
[AMD Official Use Only - AMD Internal Distribution Only]

Temp_status is initialized to zero outside of the loop. It will get modified in the for() loop. Imagine a case where in the first iteration of the loop is successful and the second iteration fails.  Then we are checking on the temp_status updated in the last successful transaction.

From: sathyapk ***@***.***>
Sent: Wednesday, August 13, 2025 2:38 AM
To: AMDESE/linux-aspeed ***@***.***>
Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***>
Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@sathyapk commented on this pull request.

________________________________

In drivers/misc/amd-apml/apml_alertl.c<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272355408>:

> +            return IRQ_NONE;

+

+       for (i = 0; i < oob_adata->num_of_tsi_devs; i++) {

+              rt_src = 0;

+              /* Check for NULL pointer */

+              if (IS_ERR_OR_NULL(oob_adata->tsi_dev[i])) {

+                      dev_err(dev, "TSI device does not exist\n");

+                      return IRQ_NONE;

+              }

+

+              /* Check for TSI alert */

+              ret = tsi_alert_check(oob_adata, &temp_status, i);

+              if (ret < 0)

+                      dev_err(dev, "Failed to read TSI status register\n");

+

+              if (!temp_status)

temp_status initialized to zero. On failure, temp_status remains zero

—
Reply to this email directly, view it on GitHub<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272355408>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJVER2A6ERYLZVAJHCT3NLTMFAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMYTKNBRGM>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

mahkurap commented:
[AMD Official Use Only - AMD Internal Distribution Only]

How does the apml_library know the structure of the event? Is there a header file defined for it? Imagine a case where in the user wants to use the events directly with out the library interface. Is it not possible? Library also runs in the user space.

From: sathyapk ***@***.***>
Sent: Wednesday, August 13, 2025 2:39 AM
To: AMDESE/linux-aspeed ***@***.***>
Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***>
Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@sathyapk commented on this pull request.

________________________________

In drivers/misc/amd-apml/apml_alertl.c<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272358558>:

> +     char sock[MAX_SOC_LEN];

+       char src[MAX_ERR_LEN];

+       char *alert_source[] = { sock, src, NULL };

+

+       soc_die_num = static_addr_to_socket(static_address);

+       if (soc_die_num == 0xFF) {

+              dev_err(dev, "Device static address not valid\n");

+              return -ENODEV;

+       }

+

+       snprintf(sock, sizeof(sock), "Socket=0x%x", soc_die_num);

+       snprintf(src, sizeof(src), "Source=0x%x", alert_src);

+

+       dev_err(dev, "Sending uevent: Sock:0x%x Src:0x%x\n",

+              soc_die_num, alert_src);

+       kobject_uevent_env(&dev->kobj, KOBJ_CHANGE, alert_source);

apml_library is the interface to the alert_l driver. user space application should call apml library api's to interpret the data

—
Reply to this email directly, view it on GitHub<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272358558>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJVI6KJ4DOPVIEDGKOT3NLTRPAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMYTSNZXGM>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

mahkurap commented:
[AMD Official Use Only - AMD Internal Distribution Only]

Bottom halfs are called/handled as part of work queues as far as I remember. I could be wrong. Function tsi_alert_check() is called in “static irqreturn_t alert_l_irq_thread_handler(int irq, void *dev_id)” which is an irq handler.

From: sathyapk ***@***.***>
Sent: Wednesday, August 13, 2025 2:40 AM
To: AMDESE/linux-aspeed ***@***.***>
Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***>
Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@sathyapk commented on this pull request.

________________________________

In drivers/misc/amd-apml/apml_alertl.c<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272361147>:

> +#define MAX_SOC_LEN             11

+#define MAX_ERR_LEN             18

+

+MODULE_ALIAS("apml_alertl:" DRIVER_NAME);

+

+/* Read TSI Status register to identify the RAS error */

+static int tsi_alert_check(struct apml_alertl_data *oob_adata, int *temp_status, u8 soc_die_num)

+{

+       struct apml_message msg = { 0 };

+       int ret;

+

+       if (!oob_adata->tsi_dev[soc_die_num] || !oob_adata->tsi_dev[soc_die_num]->regmap)

+              return 0;

+       msg.data_in.reg_in[REG_OFF_INDEX] = TSI_STATUS_REG;

+

+       mutex_lock(&oob_adata->tsi_dev[soc_die_num]->lock);

Lock is called inthe bottom half of the irq, mutex lock can be invoked in bottom half.

—
Reply to this email directly, view it on GitHub<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272361147>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJUEFFVKKDKFMC2L3CT3NLTVRAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMZDEOBVGU>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

mahkurap commented:
[AMD Official Use Only - AMD Internal Distribution Only]

I need to understand more on how the mutexes initialized in a different driver can be accessed in this driver. Something new for me. How do you enforce the probe order()? First sbrmi, sbtsi and then apml_alertl?

From: sathyapk ***@***.***>
Sent: Wednesday, August 13, 2025 2:45 AM
To: AMDESE/linux-aspeed ***@***.***>
Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***>
Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143)

Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding.


@sathyapk commented on this pull request.

________________________________

In drivers/misc/amd-apml/apml_alertl.c<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272374417>:

> +     /* Allocate memory as per the number of TSI devices */

+       tsi_dev = devm_kzalloc(dev, oob_alert->num_of_tsi_devs * sizeof(struct apml_sbtsi_device),

+                             GFP_KERNEL);

+       if (!tsi_dev)

+              return -ENOMEM;

+

+       oob_alert->tsi_dev = tsi_dev;

+       oob_alert->dev = dev;

+

+       /*

+        * For each of the Alerts get the device associated

+        * Currently the ALert_L driver identification is only supported

+        * over I3C. We can add property in dts to identify the bus type

+        */

+

+       for (i = 0; i < oob_alert->num_of_rmi_devs; i++) {

The mutex locks are initialized in respective RMI and TSI drivers.

—
Reply to this email directly, view it on GitHub<https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272374417>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJTJHUOSPXIDWSXSLEL3NLUI5AVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGM2DCNRTHA>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

sathyapk commented:
> [AMD Official Use Only - AMD Internal Distribution Only] Temp_status is initialized to zero outside of the loop. It will get modified in the for() loop. Imagine a case where in the first iteration of the loop is successful and the second iteration fails. Then we are checking on the temp_status updated in the last successful transaction. From: sathyapk ***@***.***> Sent: Wednesday, August 13, 2025 2:38 AM To: AMDESE/linux-aspeed ***@***.***> Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***> Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143) Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding. @sathyapk commented on this pull request.
> ________________________________ In drivers/misc/amd-apml/apml_alertl.c<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272355408)>:
> + return IRQ_NONE;
> + + for (i = 0; i < oob_adata->num_of_tsi_devs; i++) { + rt_src = 0; + /* Check for NULL pointer */ + if (IS_ERR_OR_NULL(oob_adata->tsi_dev[i])) { + dev_err(dev, "TSI device does not exist\n"); + return IRQ_NONE; + } + + /* Check for TSI alert */ + ret = tsi_alert_check(oob_adata, &temp_status, i); + if (ret < 0) + dev_err(dev, "Failed to read TSI status register\n"); + + if (!temp_status) temp_status initialized to zero. On failure, temp_status remains zero — Reply to this email directly, view it on GitHub<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272355408)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJVER2A6ERYLZVAJHCT3NLTMFAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMYTKNBRGM>. You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

This is a miss, thanks for pointing it out. I will fix this.

sathyapk commented:
> [AMD Official Use Only - AMD Internal Distribution Only] Bottom halfs are called/handled as part of work queues as far as I remember. I could be wrong. Function tsi_alert_check() is called in “static irqreturn_t alert_l_irq_thread_handler(int irq, void *dev_id)” which is an irq handler. From: sathyapk ***@***.***> Sent: Wednesday, August 13, 2025 2:40 AM To: AMDESE/linux-aspeed ***@***.***> Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***> Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143) Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding. @sathyapk commented on this pull request.
> ________________________________ In drivers/misc/amd-apml/apml_alertl.c<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272361147)>:
> +#define MAX_SOC_LEN 11
> +#define MAX_ERR_LEN 18 + +MODULE_ALIAS("apml_alertl:" DRIVER_NAME); + +/* Read TSI Status register to identify the RAS error */ +static int tsi_alert_check(struct apml_alertl_data *oob_adata, int *temp_status, u8 soc_die_num) +{ + struct apml_message msg = { 0 }; + int ret; + + if (!oob_adata->tsi_dev[soc_die_num] || !oob_adata->tsi_dev[soc_die_num]->regmap) + return 0; + msg.data_in.reg_in[REG_OFF_INDEX] = TSI_STATUS_REG; + + mutex_lock(&oob_adata->tsi_dev[soc_die_num]->lock); Lock is called inthe bottom half of the irq, mutex lock can be invoked in bottom half. — Reply to this email directly, view it on GitHub<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272361147)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJUEFFVKKDKFMC2L3CT3NLTVRAVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGMZDEOBVGU>. You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

There are multiple ways to implement bottom half, like softirqs, tasklets, workqueue, threaded irq handler.
We are using threaded IRQ.

sathyapk commented:
> [AMD Official Use Only - AMD Internal Distribution Only] I need to understand more on how the mutexes initialized in a different driver can be accessed in this driver. Something new for me. How do you enforce the probe order()? First sbrmi, sbtsi and then apml_alertl? From: sathyapk ***@***.***> Sent: Wednesday, August 13, 2025 2:45 AM To: AMDESE/linux-aspeed ***@***.***> Cc: Kurapati, Mahesh ***@***.***>; Review requested ***@***.***> Subject: Re: [AMDESE/linux-aspeed] alert_L module for handling oob alerts (PR #143) Caution: This message originated from an External Source. Use proper caution when opening attachments, clicking links, or responding. @sathyapk commented on this pull request.
> ________________________________ In drivers/misc/amd-apml/apml_alertl.c<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272374417)>:
> + /* Allocate memory as per the number of TSI devices */
> + tsi_dev = devm_kzalloc(dev, oob_alert->num_of_tsi_devs * sizeof(struct apml_sbtsi_device), + GFP_KERNEL); + if (!tsi_dev) + return -ENOMEM; + + oob_alert->tsi_dev = tsi_dev; + oob_alert->dev = dev; + + /* + * For each of the Alerts get the device associated + * Currently the ALert_L driver identification is only supported + * over I3C. We can add property in dts to identify the bus type + */ + + for (i = 0; i < oob_alert->num_of_rmi_devs; i++) { The mutex locks are initialized in respective RMI and TSI drivers. — Reply to this email directly, view it on GitHub<[#143 (comment)](https://github.com/AMDESE/linux-aspeed/pull/143#discussion_r2272374417)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCUUIJTJHUOSPXIDWSXSLEL3NLUI5AVCNFSM6AAAAAB7HM3CZ2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJUGM2DCNRTHA>. You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

Alert_l driver is dependent on rmi and tsi driver taken care in Kconfig
Probing of rmi/tsi device are not part of Alert_L driver. Alert_L driver only access the rmi/tsi device structure info updated during rmi/tsi driver init(probe).

mahkurap commented on arch/arm64/boot/dts/aspeed/aspeed-bmc-amd-congo.dts:
```diff
@@ -907,6 +907,17 @@
+		status = "okay";
+		gpios = <&ltpi0_gpio 20 GPIO_ACTIVE_LOW>;
+		sbrmi = <&sbrmi_p0_iod0>;
+		sbtsi = <&sbtsi_p0_iod0 &sbtsi_p0_iod1>;
```
iod1 may not be needed. I know I asked for it in the initial review. At the time BMC was reading the the iod1 temperature for the fan curve so I thought we might need it for as well. please check the requirement and add it if there is a requirement for it. apologies once again.

sathyapk replied:
Sure, will confirm and update. thanks.

sathyapk replied:
CF team confirmed IOD1 is not applicable in mission mode. Hence removed IOD1 related changes in DTS.
